pkg/reconciler/channel/channel.go

```diff
@@ -0,0 +1,412 @@
+	// Convert the namespace/name string into a distinct namespace and name
+	namespace, name, err := cache.SplitMetaNamespaceKey(key)
+	if err != nil {
+		c.Logger.Errorf("invalid resource key: %s", key)
+		return nil
+	}
+	logger := logging.FromContext(ctx)
+
+	// Get the Channel resource with this namespace/name
+	original, err := c.channelLister.Channels(namespace).Get(name)
+	if apierrs.IsNotFound(err) {
+		// The resource may no longer exist, in which case we stop processing.
+		logger.Errorf("Channel %q in work queue no longer exists", key)
+		return nil
+	} else if err != nil {
+		return err
+	}
+
+	if original.GetDeletionTimestamp() != nil {
+		return nil
+	}
+
+	// Don't modify the informers copy
+	channel := original.DeepCopy()
+
+	// Reconcile this copy of the channel and then write back any status
+	// updates regardless of whether the reconciliation errored out.
+	var reconcileErr = c.reconcile(ctx, channel)
+
+	// If no error is returned, mark the observed generation.
+	if reconcileErr == nil {
+		channel.Status.ObservedGeneration = channel.Generation
+	}
+
+	if equality.Semantic.DeepEqual(original.Status, channel.Status) {
+		// If we didn't change anything then don't call updateStatus.
+		// This is important because the copy we loaded from the informer's
+		// cache may be stale and we don't want to overwrite a prior update
+		// to status with this stale state.
+
+	} else if _, uErr := c.updateStatus(ctx, channel); uErr != nil {
+		logger.Warnw("Failed to update Channel status", zap.Error(uErr))
+		c.Recorder.Eventf(channel, corev1.EventTypeWarning, "UpdateFailed",
+			"Failed to update status for Channel %q: %v", channel.Name, uErr)
+		return uErr
+	} else if reconcileErr == nil {
+		// There was a difference and updateStatus did not return an error.
+		c.Recorder.Eventf(channel, corev1.EventTypeNormal, "Updated", "Updated Channel %q", channel.GetName())
+	}
+	if reconcileErr != nil {
+		c.Recorder.Event(channel, corev1.EventTypeWarning, "InternalError", reconcileErr.Error())
+	}
+	return reconcileErr
+}
+
+func (r *Reconciler) reconcile(ctx context.Context, channel *v1alpha1.Channel) error {
+	channel.Status.InitializeConditions()
+
+	if channel.Status.TopicID == "" {
+		channel.Status.TopicID = resources.GenerateTopicName(channel.UID)
+	}
+
+	// 1. Create the Topic.
+	if topic, err := r.createTopic(ctx, channel); err != nil {
+		channel.Status.MarkNoTopic("CreateTopicFailed", "Error when attempting to create Topic.")
+		return err
+	} else {
+		// Propagate Status.
+		if c := topic.Status.GetCondition(pubsubv1alpha1.TopicConditionReady); c != nil {
+			if c.IsTrue() {
+				channel.Status.MarkTopicReady()
+			} else if c.IsUnknown() {
+				channel.Status.MarkTopicOperating(c.Reason, c.Message)
+			} else if c.IsFalse() {
+				channel.Status.MarkNoTopic(c.Reason, c.Message)
+			}
+		}
+	}
+	// 2. Sync all subscriptions.
+	//   a. create all subscriptions that are in spec and not in status.
```

Comment:
Maybe also add the case about update. Also, consider moving these comments (also maybe expand a little bit) to the actual syncSubscribers. By the time I was reading that function, I didn't remember reading this comment, and tracking to the callsite to see the behaviour of the function is odd :)

Reply:
When this function gets reworked I will do this too.